runtime-sdk/src/context.rs

```diff
@@ -147,9 +156,31 @@ impl<'a> DispatchContext<'a> {
             max_messages: self.max_messages.saturating_sub(self.messages.len() as u32),
             messages: Vec::new(),
             values: &mut self.values,
+            tx_values: BTreeMap::new(),
         };
         f(tx_ctx, tx.call)
     }
+
+    /// Run something with a simulation context based on this context.
+    /// The simulation context collects its own messages and starts with an empty set of context
+    /// values.
+    pub fn with_simulation<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut DispatchContext<'_>) -> R,
+    {
+        let mut sim_ctx = DispatchContext {
+            mode: Mode::SimulateTx,
+            runtime_header: self.runtime_header,
+            runtime_round_results: self.runtime_round_results,
+            runtime_storage: self.runtime_storage,
```

Comment:
Note that this way anything in the simulation context could still overwrite storage if it commits something. I think it is enough to mention this in the method documentation.

Reply:
would be better to have an overlay

Reply:
Yeah it would be if you can make it so easily.

Reply:
oh right I can't. I just realized there's no overlay defined for wrapping the panicky kind of MKVS. comment it is!

Reply:
https://github.com/oasisprotocol/oasis-core/issues/3904